README.md

[![R-CMD-check](https://github.com/rinnan/CENFA/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/rinnan/CENFA/actions/workflows/R-CMD-check.yaml)
[![CRAN](https://www.r-pkg.org/badges/version/CENFA)]()
![](https://cranlogs.r-pkg.org/badges/last-month/CENFA)

<!-- [![Build Status](https://travis-ci.org/rinnan/CENFA.svg?branch=master)](https://travis-ci.org/rinnan/CENFA) -->

# CENFA: Climate and Ecological Niche Factor Analysis

`CENFA` provides tools for performing ecological-niche factor analysis
(ENFA) and climate-niche factor analysis (CNFA). This package was
created with three goals in mind:

- To update the ENFA method for use with large datasets and modern data
  formats.
- To expand the application of ENFA in the context of climate change in
  order to quantify different aspects of species vulnerability to
  climate change, and to facilitate quantitative comparisons of
  vulnerability between species.
- To correct a minor error in the ENFA method itself, that has persisted
  in the literature since Hirzel et al. first introduced ENFA in 2002.

`CENFA` takes advantage of the `raster` and `sp` packages, allowing the
user to conduct analyses directly with raster, shapefile, and point
data, and to handle large datasets efficiently via partial data loading
and parallelization.

## Installation

You can install the most recent version of CENFA from GitHub with:

``` r
# install.packages("devtools")
devtools::install_github("rinnan/CENFA")
```

## Examples

### `enfa`

We will use some example datasets to perform a basic ENFA. The
historical climate dataset `climdat.hist` is a RasterBrick of 10 climate
variables, covering much of the western US coast. `QUGA` is a
SpatialPolygonsDataFrame of the historical range map of Oregon white oak
(*Quercus garryana*).

A plot of the data, using the one of the layers of `climdat.hist`:

![](man/figures/README-QUGA-plot-1.png)<!-- -->

The `enfa` function takes three basic arguments: the dataset of
ecological variables (`climdat.hist`), the map of species presence
(`QUGA`), and the values of `QUGA` that specify presence (in this case,
a column named “CODE”). Calling the `enfa` object by name provides a
standard summary of the ENFA results.

``` r
mod.enfa <- enfa(x = climdat.hist, s.dat = QUGA, field = "CODE")
#> Warning in showSRID(uprojargs, format = "PROJ", multiline = "NO", prefer_proj
#> = prefer_proj): Discarded datum Unknown based on WGS84 ellipsoid in Proj4
#> definition
mod.enfa
#> ENFA
#> 
#> Original function call: enfa(x = climdat.hist, s.dat = QUGA, field = "CODE")
#> 
#> Marginality factor: 
#>   MDR   ISO    TS HMmax CMmin   PWM   PDM    PS   PWQ   PDQ 
#> -0.13  0.51 -0.67 -0.03  0.67  0.71 -0.13  0.70  0.72  0.13 
#> 
#> Eigenvalues of specialization: 
#>  Marg Spec1 Spec2 Spec3 Spec4 Spec5 Spec6 Spec7 Spec8 Spec9 
#>  5.12  8.92  4.82  3.01  2.55  2.01  1.28  0.77  0.68  0.36 
#> 
#> Percentage of specialization contained in ENFA factors: 
#>  Marg Spec1 Spec2 Spec3 Spec4 Spec5 Spec6 Spec7 Spec8 Spec9 
#> 17.36 30.24 16.32 10.20  8.64  6.80  4.33  2.61  2.30  1.21 
#> 
#> Overall marginality:  1.654 
#> 
#> Overall specialization:  1.718 
#> 
#> Significant ENFA factors: 
#>        Marg Spec1 Spec2 Spec3
#> MDR   -0.13 -0.46  0.59  0.38
#> ISO    0.51  0.11 -0.46 -0.05
#> TS    -0.67 -0.34 -0.54  0.39
#> HMmax -0.03  0.58 -0.07 -0.57
#> CMmin  0.67 -0.54  0.00  0.49
#> PWM    0.71 -0.08  0.26  0.27
#> PDM   -0.13  0.10  0.09  0.03
#> PS     0.70  0.12 -0.11 -0.02
#> PWQ    0.72  0.03 -0.19 -0.25
#> PDQ    0.13  0.00 -0.11 -0.04
```

### `scatter`

We can visualize the ENFA results via the `scatter` function, which
produces a biplot of the marginality axis and one of the specialization
axes. This gives us a portrait of the species’ niche to compare with the
global niche of the reference study area, with the ecological axes
projected onto the ENFA dimensions. (Note: since `mod.enfa` only
contains information about the species habitat, we must first construct
a `GLcenfa` object that also describes the global habitat.)

``` r
glc <- GLcenfa(x = climdat.hist)
scatter(x = mod.enfa, y = glc)
```

![](man/figures/README-scatter-1.png)<!-- -->

    #> NULL

For larger datasets, we can speed up the computation via
parallelization. We provide two additional arguments, `parallel = TRUE`,
and `n`, which specifies the number of cores to use. `n` has a default
value of 1, so only setting `parallel = TRUE` will not parallelize the
function by itself.

``` r
# does not enable parallelization
mod <- enfa(x = climdat.hist, s.dat = QUGA, field = "CODE", parallel = TRUE)

# enables parallelization across 4 cores
mod <- enfa(x = climdat.hist, s.dat = QUGA, field = "CODE", parallel = TRUE, n = 4)
```

The function will attempt to match the value provided to `n` with the
number of cores detected on the local device via
`parallel::detectCores()`; if the provided `n` is greater than the
number of available cores `k`, a warning will be issued and `n` will be
set to `k - 1`.

### `cnfa`

The `cnfa` function is very similar to `enfa`, but performs a slightly
different analysis. Whereas ENFA returns a *specialization factor* (the
eigenvalues of specialization) describing the amount of specialization
found in each *ENFA factor*, CNFA returns a *sensitivity factor* that
reflects the amount of sensitivity found in each *ecological variable*.
This makes the sensitivity factor more directly comparable to the
marginality factor, and more interpretable in the context of species’
sensitivity to a given variable.

``` r
mod.cnfa <- cnfa(x = climdat.hist, s.dat = QUGA, field = "CODE")
mod.cnfa
#> CNFA
#> 
#> Original function call: cnfa(x = climdat.hist, s.dat = QUGA, field = "CODE")
#> 
#> Marginality factor: 
#>   MDR   ISO    TS HMmax CMmin   PWM   PDM    PS   PWQ   PDQ 
#> -0.13  0.51 -0.67 -0.03  0.67  0.71 -0.13  0.70  0.72  0.13 
#> 
#> Sensitivity factor: 
#>   MDR   ISO    TS HMmax CMmin   PWM   PDM    PS   PWQ   PDQ 
#>  4.52  2.86  4.16  3.72  3.98  3.36  1.35  1.71  2.94  0.92 
#> 
#> Percentage of specialization contained in CNFA factors: 
#>  Marg Spec1 Spec2 Spec3 Spec4 Spec5 Spec6 Spec7 Spec8 Spec9 
#> 17.36 30.24 16.32 10.20  8.64  6.80  4.33  2.61  2.30  1.21 
#> 
#> Overall marginality:  1.654 
#> 
#> Overall sensitivity:  1.718 
#> 
#> Significant CNFA factors: 
#>        Marg Spec1 Spec2 Spec3
#> MDR   -0.13 -0.46  0.59  0.38
#> ISO    0.51  0.11 -0.46 -0.05
#> TS    -0.67 -0.34 -0.54  0.39
#> HMmax -0.03  0.58 -0.07 -0.57
#> CMmin  0.67 -0.54  0.00  0.49
#> PWM    0.71 -0.08  0.26  0.27
#> PDM   -0.13  0.10  0.09  0.03
#> PS     0.70  0.12 -0.11 -0.02
#> PWQ    0.72  0.03 -0.19 -0.25
#> PDQ    0.13  0.00 -0.11 -0.04
```

Using the `sensitivity_map` function, we can create a habitat map that
identifies where we expect the species to be most sensitive to changes
in climate.

``` r
s.map <- sensitivity_map(mod.cnfa)
```

![](man/figures/README-sensitivity-map-1.png)<!-- -->

    #> NULL

### `departure`

The `departure` function provides a measure of a species’ potential
exposure to climate change. It takes a future climate dataset as an
additional argument, and calculates the absolute differences between
historical and future values.

``` r
dep <- departure(x = climdat.hist, y = climdat.fut, s.dat = QUGA, field = "CODE")
dep
#> CLIMATIC DEPARTURE
#> 
#> Departure factor: 
#>   MDR   ISO    TS HMmax CMmin   PWM   PDM    PS   PWQ   PDQ 
#>  0.05  0.10  0.22  0.53  0.44  0.23  0.12  0.38  0.23  0.16 
#> 
#> Overall departure: 0.909
```

The departure factor tells us the average amount of change that is
expected in each climate variable across the species’ range. Using the
`exposure_map` function, we can create a habitat map that identifies
where we expect the species to be most exposed to climate change.

``` r
e.map <- exposure_map(dep)
```

![](man/figures/README-exposure-map-1.png)<!-- -->

    #> NULL

### `vulnerability`

The `vulnerability` function provides a measure of a species’ potential
vulnerability to climate change, taking both sensitivity and exposure
into account. It takes a `cnfa` object and a `departure` object as its
arguments.

``` r
vuln <- vulnerability(cnfa = mod.cnfa, dep = dep)
vuln
#> CLIMATIC VULNERABILITY
#> 
#> Vulnerability factor: 
#>   MDR   ISO    TS HMmax CMmin   PWM   PDM    PS   PWQ   PDQ 
#>  2.18  1.78  2.25  2.38  2.40  2.03  1.23  1.54  1.90  1.03 
#> 
#> Overall vulnerability: 1.368
```

Using the `vulnerability_map` function, we can create a habitat map that
identifies where we expect the species to be most vulnerable to climate
change.

``` r
v.map <- vulnerability_map(vuln)
```

![](man/figures/README-vulnerability-map-1.png)<!-- -->

    #> NULL

## Useful raster functions

The `raster` package contains the `clusterR` function, which enables
parallelization methods for certain raster operations. `clusterR` only
works on functions that operate on a cell-by-cell basis, however, which
limits its usefulness. The `CENFA` package contains a few functions that
speed up some basic `raster` functions considerably by parallelizing on
a layer-by-layer basis rather than a cell-by-cell basis.

### `parScale`

The `parScale` function is identical to `raster::scale`, but has a
parallelization option that will scale each raster layer in parallel.
The `center` and `scale` arguments can be logical (`TRUE` or `FALSE`) or
numeric vectors.

``` r
clim.scaled <- parScale(x = climdat.hist, parallel = TRUE, n = 4)
```

### `parCov`

The `parCov` function returns the covariance matrix of a Raster\* object
`x`, computing the covariance between each layer of `x`. This is similar
to `raster::layerStats(x, stat = 'cov')`, but much faster when
parallelization is employed.

``` r
mat <- parCov(x = climdat.hist, parallel = TRUE, n = 4)
```

Additionally, `parCov` can accept two Raster\* objects as arguments,
similar to `stats::cov(x, y)`. If two Raster\* objects are supplied,
then the covariance is calculated between the layers of `x` and the
layers of `y`.

``` r
mat <- parCov(x = climdat.hist, y = climdat.fut, parallel = TRUE, n = 4)
```

### `stretchPlot`

The `stretchPlot` function provides a simple way to adjust the contrast
of plots of RasterLayers to emphasize difference in values. It can
perform histogram equalization and standard deviation stretching.

``` r
sm <- sensitivity_map(mod.cnfa)
par(mfrow = c(1, 3), oma = c(1,1,1,1))
stretchPlot(sm, main = "linear")
#> NULL
stretchPlot(sm, type = "hist.equal", main = "Histogram equalization")
#> NULL
stretchPlot(sm, type = "sd", n = 2, main = "Standard deviation (n = 2)")
```

![](man/figures/README-stretchPlot-1.png)<!-- -->

    #> NULL

## Guidelines for contributing

I welcome contributions and suggestions for improving this package.
Please do not hesitate to submit any issues you may encounter.

## References

Rinnan, D. Scott and Lawler, Joshua. Climate‐niche factor analysis: a
spatial approach to quantifying species vulnerability to climate change.
Ecography (2019).
[doi:10.1111/ecog.03937](https://doi.org/10.1111/ecog.03937)

Basille, Mathieu, et al. Assessing habitat selection using multivariate
statistics: Some refinements of the ecological-niche factor analysis.
Ecological Modelling 211.1 (2008): 233-240.

Hirzel, Alexandre H., et al. Ecological-niche factor analysis: how to
compute habitat-suitability maps without absence data?. Ecology 83.7
(2002): 2027-2036.


NEWS.md

# News

## CENFA 1.1.2

Published on CRAN, 2023-07-18.

- Updated methods for detecting projection mismatches between climate
  and species data. (`sp::identicalCRS()` no longer works reliably with
  Raster\* objects.)

## CENFA 1.1.1

Published on CRAN, 2021-06-08.

## CENFA 1.1.0.9000

In current development.

- Fixed bug in `enfa()` function that returned error for large datasets.

- Fixed bug in `predict()` methods for `enfa` objects that returned
  raster brick instead of layer.

- Removed `closeAllConnections()` command that may cause unwanted
  behavior.

## CENFA 1.1.0

Published on CRAN, 2020-02-15.

## CENFA 1.0.0.9000

- Updated descriptions of CNFA sensitivity and ENFA specialization.

- Fixed small bug in `parCov()` function that caused NA observations to
  be discarded.

- Added more informative error messages for `cnfa()` function.

- Forward compatibility with upcoming R 4.0 release.

## CENFA 1.0.0

Published on CRAN, 2018-11-06.

## CENFA 0.1.0.9000

- Changed `quiet = TRUE` arguments to `progress = FALSE` for more
  intuitive interface.

- Different definition of CNFA sensitivity factor and overall
  sensitivity, to better agree with ENFA specialization factor.

- Different definition of vulnerability factor and overall
  vulnerability.

- Fixed bugs in `parScale` function involving parallel methods.

- Added predict methods for `enfa`, `cnfa`, `departure` and
  `vulnerability` objects.

- Fixed bug in `GLcenfa` that prevented writing of Raster\* objects.

- Imports `raster` package instead of depends.

- Fixes bug in `parCov` function involving covariance calculations
  between two Raster\* objects.

## CENFA 0.1.0

Published on CRAN, as of 2018-02-06!


R/cnfa.R
#' Climate-niche factor analysis
#'
#' Performs climate-niche factor analysis using climate raster data and species
#' presence data.
#'
#' @aliases print.cnfa show.cnfa
#'
#' @param x Raster* object, typically a brick or stack with p climate
#'   raster layers, or a \code{GLcenfa} object
#' @param s.dat RasterLayer, SpatialPolygons*, or SpatialPoints* object indicating
#'   species presence or abundance
#' @param field field of \code{s.dat} that specifies presence or abundance. This
#'   is equivalent to the \code{field} argument in \code{\link[raster]{rasterize}}
#' @param fun function or character. Determines what values to assign to cells
#'   with multiple spatial features, similar to the \code{fun} argument in
#'   \code{\link[raster]{rasterize}}.  Options are 'first', 'last' (default),
#'   and 'count' (see Details)
#' @param scale logical. If \code{TRUE} then the values of \code{x} will get
#'   centered and scaled. Depending on the resolution of the climate data and
#'   the extent of the study area, this can be quite time consuming. If running
#'   this function for multiple species, it is recommended that the data be
#'   scaled beforehand using the \code{\link{GLcenfa}} function
#' @param filename character. Optional filename to save the Raster* output to
#'   file. If this is not provided, a temporary file will be created for large \code{x}
#' @param progress logical. If \code{TRUE}, messages and progress bar will be
#'   printed
#' @param parallel logical. If \code{TRUE} then multiple cores are utilized for the
#'   calculation of the covariance matrices
#' @param n numeric. Number of CPU cores to utilize for parallel processing
#' @param cl optional cluster object
#' @param keep.open logical. If \code{TRUE} and \code{parallel = TRUE}, the
#'   cluster object will not be closed after the function has finished
#' @param ... Additional arguments for \code{\link[raster]{writeRaster}}
#'
#' @examples
#' mod1 <- cnfa(x = climdat.hist, s.dat = ABPR, field = "CODE")
#'
#' # using GLcenfa as an initial step
#' # for multi-species comparison
#'
#' glc <- GLcenfa(x = climdat.hist)
#' mod2 <- cnfa(x = glc, s.dat = ABPR, field = "CODE")
#'
#'# same results either way
#' all.equal(m.factor(mod1), m.factor(mod2))
#' all.equal(s.factor(mod1), s.factor(mod2))
#'
#' @return Returns an S4 object of class \code{cnfa} with the following components:
#' \describe{
#'   \item{call}{Original function call}
#'   \item{mf}{Marginality factor. Vector of length p that describes the location
#'   of the species Hutchinsonian niche relative to the global niche}
#'   \item{marginality}{Magnitude of the marginality factor}
#'   \item{sf}{Sensitivity factor. Vector of length p that describes the amount of
#'    sensitivity for each climate variable}
#'   \item{sensitivity}{Square root of the mean of the sensitivity factor}
#'   \item{eig}{Named vector of eigenvalues of specialization for each CNFA factor}
#'   \item{co}{A p x p matrix describing the amount of marginality and specialization
#'    in each CNFA factor.}
#'   \item{cov}{p x p species covariance matrix}
#'   \item{g.cov}{p x p global covariance matrix}
#'   \item{ras}{RasterBrick of transformed climate values, with p layers}
#'   \item{weights}{Raster layer of weights used for CNFA calculation}
#' }
#'
#' @details
#' The \code{cnfa} function is not to be confused with the
#' \code{\link{enfa}} function. \code{enfa} performs ENFA as described by Hirzel
#' et al. (2002) and Basille et al. (2008), and is offered as an alternative to
#' the \code{enfa} function in the \code{adehabitatHS} package. There are
#' several key differences between ENFA and CNFA.
#'
#' Whereas ENFA returns a \strong{specialization factor} that describes
#' the specialization in each \strong{ENFA factor}, CNFA returns a
#' \strong{sensitivity factor} \code{sf} that describes the sensitivity in each
#' \strong{environmental variable}. This makes the sensitivity factor more
#' directly comparable to the marginality factor \code{mf}, because their
#' dimensions are identical. Sensitivity is calculated by a weighted sum
#' of the amount of specialization found in each CNFA factor, \emph{including}
#' the marginality factor. As such, the sensitivity factor offers a more complete
#' measure of specialization than ENFA's specialization factor, which does
#' not calculate the amount of specialization found in the marginality factor.
#' As such, CNFA's overall sensitivity (found in the slot \code{sensitivity})
#' offers a more complete measure of niche specialization than ENFA's overall
#' specialization (found in the slot \code{specialization}).
#'
#' The default \code{fun = 'last'} gives equal weight to each occupied cell.
#' If multiple species observations occur in the same cell, the cell will only
#' be counted once. \code{fun = 'count'} will weight the cells by the number
#' of observations.
#'
#' If there is too much correlation between the layers of \code{x}, the global
#' covariance matrix will be singular, and the overall marginality and overall
#' sensitivity will not be meaningful. In this case, a warning is issued,
#' and \code{marginality} and \code{sensitivity} are both returned as \code{NA}.
#'
#' @references
#' Rinnan, D. Scott and Lawler, Joshua. Climate-niche factor analysis: a spatial
#' approach to quantifying species vulnerability to climate change. Ecography (2019):
#' <doi:10.1111/ecog.03937>.
#'
#' Basille, Mathieu, et al. Assessing habitat selection using multivariate
#' statistics: Some refinements of the ecological-niche factor analysis. Ecological
#' Modelling 211.1 (2008): 233-240.
#'
#' Hirzel, Alexandre H., et al. Ecological-niche factor analysis: how to compute
#' habitat-suitability maps without absence data?. Ecology 83.7 (2002): 2027-2036.
#'
#' @seealso \code{\link{GLcenfa}}, \code{\link{enfa}}
#'
#' @export
#' @name cnfa
#'
#' @importFrom stats cov
#' @importFrom methods as is
# @importFrom raster rasterTmpFile intersect extent values mask crop values<-

setGeneric("cnfa", function(x, s.dat, ...){
  standardGeneric("cnfa")})

#' @rdname cnfa
setMethod("cnfa",
          signature(x = "GLcenfa", s.dat = "Raster"),
          function(x, s.dat, filename = "", progress = FALSE, parallel = FALSE, n = 1, cl = NULL, keep.open = FALSE, ...){

            call <- sys.call(sys.parent())
            call <- match.call(cnfa, call)

            if(file.exists(filename)) {
              params <- list(...)
              if(length(params) > 0) {
                if(is.null(params$overwrite)) {
                  stop(paste0(filename, " exists. use 'overwrite=TRUE' if you want to overwrite it"))
                } else if(!(params$overwrite)) {
                  stop(paste0(filename, " exists. use 'overwrite=TRUE' if you want to overwrite it"))
                }
              } else stop(paste0(filename, " exists. use 'overwrite=TRUE' if you want to overwrite it"))
            }

            if (nlayers(s.dat) > 1) stop('"s.dat" should be a single RasterLayer')
            # if (!identicalCRS(raster(x), s.dat)) stop("climate and species projections do not match")
            if (!raster::compareCRS(raster(x), s.dat)) stop("climate and species projections do not match")
            nS <- length(which(values(s.dat) > 0))
            if(nS == 1) stop("CNFA is not meaningful for single observations")
            ras <- raster(x)
            ext <- extent(ras)
            ext.s <- extent(s.dat)
            if (is.null(intersect(ext, ext.s))) stop("climate and species data do not overlap")
            if (raster::union(ext, ext.s) != ext) stop("extent of species data not contained within extent of climate data")

            x.crop <- crop(ras, ext.s)

            filename <- trim(filename)
            if (!canProcessInMemory(x.crop) && filename == '') {
              filename <- rasterTmpFile()
            }

            if (canProcessInMemory(x.crop) && !parallel) {
              pres <- which(!is.na(values(s.dat)) & !is.na(values(max(x.crop))))
              S <- values(x.crop)[pres, ]
              nS <- nrow(S)
              Rg <- x@cov
              p <- values(s.dat)[pres]
              p.sum <- sum(p)
              mar <- apply(S, 2, function(x) sum(x * p)) / p.sum
              if (progress) cat("Calculating species covariance matrix...\n")
              Sm <- sweep(S, 2, mar)
              DpSm <- apply(Sm, 2, function(x) x * p)
              Rs <- crossprod(Sm, DpSm) * 1 / (p.sum - 1)
            } else {
              x.mask <- mask(x.crop, s.dat)
              Rg <- x@cov
              p.sum <- cellStats(s.dat, sum)
              DpS <- x.mask * s.dat
              mar <- cellStats(DpS, sum) / p.sum
              if (progress) cat("Calculating species covariance matrix...\n")
              if (parallel) {
                if (!keep.open) on.exit(closeAllConnections())
                if (missing(cl) && n > 1) cl <- snow::makeCluster(getOption("cl.cores", n))
              }
              Sm <- parScale(x.mask, center = mar, scale = F, parallel = parallel, n = n, progress = F, keep.open = keep.open, cl = cl)
              Rs <- parCov(x = Sm, w = s.dat, parallel = parallel, n = n, progress = progress, keep.open = keep.open, cl = cl)
            }

            cZ <- nlayers(ras)
            m <- sqrt(as.numeric(t(mar) %*% mar))
            if (max(Im(eigen(Rs)$values)) > 1e-05) stop("complex eigenvalues. Try removing correlated variables.")
            eigRs <- lapply(eigen(Rs), Re)
            keep <- (eigRs$values > 1e-09)
            Rs12 <- eigRs$vectors[, keep] %*% diag(eigRs$values[keep]^(-0.5)) %*% t(eigRs$vectors[, keep])
            #Rs12 <- eigRs$vectors %*% diag(eigRs$values^(-0.5)) %*% t(eigRs$vectors)
            W <- Rs12 %*% Rg %*% Rs12
            z <- Rs12 %*% mar
            y <- z/sqrt(sum(z^2))
            H <- (diag(cZ) - y %*% t(y)) %*% W %*% (diag(cZ) - y %*% t(y))
            s <- Re(eigen(H)$values)[-cZ]
            s.p <- (t(mar) %*% Rg %*% mar) / (t(mar) %*% Rs %*% mar)
            s <- c(s.p, s)
            #s.p <- abs(s) / sum(abs(s))
            v <- Re(eigen(H)$vectors)
            U <- matrix(nrow = cZ, ncol = cZ)
            u <- as.matrix((Rs12 %*% v)[, 1:(cZ-1)])
            norw <- sqrt(diag(t(u) %*% u))
            U[, -1] <- sweep(u, 2, norw, "/")
            #co[, 1] <- mar / m
            U[, 1] <- mar
            V <- sweep(abs(U), 2, colSums(abs(U)), "/")
            sf <- as.numeric(V %*% s)
            #sens <- sqrt(as.numeric(t(sf) %*% sf)/cZ)
            sens <- sqrt(mean(sf))
            nm <- c("Marg", paste0("Spec", (1:(cZ-1))))
            if (canProcessInMemory(x.crop) & !parallel){
              s.ras <- brick(x.crop)
              if (progress) cat("Creating factor rasters...")
              values(s.ras)[pres, ] <- S %*% U
              names(s.ras) <- nm
            } else {
              if (progress) cat("Creating factor rasters...")
              f1 <- function(x) x %*% U
              if(parallel) {
                s.ras <- clusterR(x.mask, fun = .calc, args = list(fun = f1, forceapply = T, names = nm), cl = cl, filename = filename, ...)
              } else {
                s.ras <- .calc(x.mask, fun = f1, forceapply = T, filename = filename, names = nm, ...)
              }
            }
            colnames(U) <- names(s) <- nm
            rownames(U) <- names(sf) <- names(mar) <- names(ras)
            if (!keep.open || missing(cl)) snow::stopCluster(cl)

            mod <- methods::new("cnfa", call = call, mf = mar, marginality = m, sf = sf,
                                 sensitivity = sens, eig = s, co = U, cov = Rs, g.cov = Rg, ras = s.ras, weights = s.dat)
            return(mod)
          }
)

#' @rdname cnfa
setMethod("cnfa",
          signature(x = "GLcenfa", s.dat = "Spatial"),
          function(x, s.dat, field, fun = "last", filename = "", progress = FALSE, parallel = FALSE, n = 1, cl = NULL, keep.open = FALSE, ...){

            call <- sys.call(sys.parent())
            call <- match.call(cnfa, call)

            if(file.exists(filename)) {
              params <- list(...)
              if(length(params) > 0) {
                if(is.null(params$overwrite)) {
                  stop(paste0(filename, " exists. use 'overwrite=TRUE' if you want to overwrite it"))
                } else if(!(params$overwrite)) {
                  stop(paste0(filename, " exists. use 'overwrite=TRUE' if you want to overwrite it"))
                }
              } else stop(paste0(filename, " exists. use 'overwrite=TRUE' if you want to overwrite it"))
            }

            if (! inherits(s.dat, c('SpatialPolygons', 'SpatialPoints'))) stop('"s.dat" should be a "SpatialPolygons*" or "SpatialPoints*" object')
            if (!raster::compareCRS(raster(x), s.dat)) stop("climate and species projections do not match")
            ras <- raster(x)
            ext <- extent(ras)
            ext.s <- extent(s.dat)
            if (is.null(intersect(ext, ext.s))) stop("climate and species data do not overlap")
            if (raster::union(ext, ext.s) != ext) stop("extent of species data not contained within extent of climate data")

            x.crop <- crop(ras, ext.s)
            s.dat.ras <- rasterize(s.dat, x.crop, field = field, fun = fun)

            mod <- cnfa(x = x, s.dat = s.dat.ras, filename = filename, progress = progress, parallel = parallel, n = n, cl = cl, keep.open = keep.open, ...)
            mod@call <- call
            return(mod)
          }
)

#' @rdname cnfa
setMethod("cnfa",
          signature(x = "Raster", s.dat = "Raster"),
          function(x, s.dat, scale = TRUE, filename = "", progress = FALSE, parallel = FALSE, n = 1, cl = NULL, keep.open = keep.open, ...){

            call <- sys.call(sys.parent())
            call <- match.call(cnfa, call)

            if(file.exists(filename)) {
              params <- list(...)
              if(length(params) > 0) {
                if(is.null(params$overwrite)) {
                  stop(paste0(filename, " exists. use 'overwrite=TRUE' if you want to overwrite it"))
                } else if(!(params$overwrite)) {
                  stop(paste0(filename, " exists. use 'overwrite=TRUE' if you want to overwrite it"))
                }
              } else stop(paste0(filename, " exists. use 'overwrite=TRUE' if you want to overwrite it"))
            }

            if (nlayers(s.dat) > 1) stop('"s.dat" should be a single RasterLayer')
            if (!raster::compareCRS(x, s.dat)) stop("climate and species projections do not match")
            if(is.null(intersect(extent(x), extent(s.dat)))) stop("climate and species data do not overlap")
            if(raster::union(extent(x), extent(s.dat)) != extent(x)) stop("extent of species data not contained within extent of climate data")

            if (scale) {
              if (progress) cat("Scaling raster data...\n")
              #x <- parScale(x, center = T, scale = T, parallel = parallel, n = n, progress = progress)
              x <- GLcenfa(x = x, center = T, scale = T, progress = progress, parallel = parallel, n = n, cl = cl, keep.open = keep.open)
            } else x <- GLcenfa(x = x, center = F, scale = F, progress = progress, parallel = parallel, n = n, cl = cl, keep.open = keep.open)

            mod <- cnfa(x = x, s.dat = s.dat, filename = filename, progress = progress, parallel = parallel, n = n, cl = cl, keep.open = keep.open, ...)
            mod@call <- call
            return(mod)
          }
)

#' @rdname cnfa
setMethod("cnfa",
          signature(x = "Raster", s.dat = "Spatial"),
          function(x, s.dat, field, fun = "last", scale = TRUE, filename = "", progress = FALSE, parallel = FALSE, n = 1, cl = NULL, keep.open = FALSE, ...){

            call <- sys.call(sys.parent())
            call <- match.call(cnfa, call)

            if(file.exists(filename)) {
              params <- list(...)
              if(length(params) > 0) {
                if(is.null(params$overwrite)) {
                  stop(paste0(filename, " exists. use 'overwrite=TRUE' if you want to overwrite it"))
                } else if(!(params$overwrite)) {
                  stop(paste0(filename, " exists. use 'overwrite=TRUE' if you want to overwrite it"))
                }
              } else stop(paste0(filename, " exists. use 'overwrite=TRUE' if you want to overwrite it"))
            }

            if (! inherits(s.dat, c('SpatialPolygons', 'SpatialPoints'))) stop('"s.dat" should be a "SpatialPolygons*" or "SpatialPoints*" object')
            if (!raster::compareCRS(x, s.dat)) stop("climate and species projections do not match")
            if(is.null(intersect(extent(x), extent(s.dat)))) stop("climate and species data do not overlap")
            if(raster::union(extent(x), extent(s.dat)) != extent(x)) stop("extent of species data not contained within extent of climate data")

            s.dat.ras <- rasterize(s.dat, raster(x), field = field, fun = fun)
            if (scale) {
              x <- GLcenfa(x = x, center = T, scale = T, progress = progress, parallel = parallel, n = n, cl = cl, keep.open = keep.open)
            } else x <- GLcenfa(x = x, center = F, scale = F, progress = progress, parallel = parallel, n = n, cl = cl, keep.open = keep.open)

            mod <- cnfa(x = x, s.dat = s.dat.ras, filename = filename, progress = progress, parallel = parallel, n = n, cl = cl, keep.open = keep.open, ...)
            mod@call <- call
            return(mod)
          }
)

# setMethod("cnfa",
#           signature(x = "Raster", s.dat = "sf"),
#           function(x, s.dat, field, fun = "last", scale = TRUE, filename = "", progress = FALSE, parallel = FALSE, n = 1, ...){
#             if (!requireNamespace("sf")) {
#               warning('cannot do this because sf is not available')
#             }
#
#             call <- sys.call(sys.parent())
#
#             if(!identicalCRS(x, s.dat)) stop("projections do not match")
#             if(is.null(intersect(extent(x), extent(s.dat)))) stop("climate and species data do not overlap")
#             if(raster::union(extent(x), extent(s.dat)) != extent(x)) stop("extent of species data not contained within extent of climate data")
#
#             s.dat <- as(s.dat, "Spatial")
#             if (! inherits(s.dat, c('SpatialPolygons', 'SpatialPoints'))) stop('geometry of "s.dat" should be of class "sfc_POLYGON", "sfc_MULTIPOLYGON", "sfc_POINT", or "sfc_MULTIPOINT"')
#
#             s.dat.ras <- rasterize(s.dat, raster(x), field = field, fun = fun)
#             if (scale) {
#               if (progress) cat("Scaling raster data...\n")
#               x <- GLcenfa(x = x, center = T, scale = T, progress = progress, parallel = parallel, n = n)
#             } else x <- GLcenfa(x = x, center = F, scale = F, progress = progress, parallel = parallel, n = n)
#
#             cnfa(x = x, s.dat = s.dat.ras, filename = filename, progress = progress, parallel = parallel, n = n, ...)
#           }
# )
